File: stack/python/reverse_some_chars.py
from typing import Iterable


def reverse_some_chars(s: str, chars: Iterable[str]) -> str:
    target_set = set(chars)
    stack = [ch for ch in s if ch in target_set]

    result = []
    for ch in s:
        if ch in target_set:
            result.append(stack.pop())
        else:
            result.append(ch)
    return ''.join(result)

File: stack/ts/reverseSomeChars.ts
export function reverseSomeChars(input: string, chars: string[]): string {
  const set = new Set(chars);
  const stack: string[] = [];
  for (const ch of input) {
    if (set.has(ch)) {
      stack.push(ch);
    }
  }

  const result: string[] = [];
  for (const ch of input) {
    if (set.has(ch)) {
      result.push(stack.pop()!);
    } else {
      result.push(ch);
    }
  }

  return result.join('');
}
